File: app/src/main/java/com/relic/presentation/displaysub/DisplaySubVM.java
package com.relic.presentation.displaysub;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.util.Log;
import android.view.View;

import com.relic.data.PostRepository;
import com.relic.data.SubRepository;
import com.relic.presentation.callbacks.RetrieveNextListingCallback;
import com.relic.data.models.PostModel;
import com.relic.data.models.SubredditModel;

import java.util.List;


public class DisplaySubVM extends ViewModel implements DisplaySubContract.ViewModel, RetrieveNextListingCallback {
  private final String TAG = "DISPLAY_SUB_VM";
  private boolean isInitialized = false;

  private String subName;
  private SubRepository subRepo;
  private PostRepository postRepo;

  private MediatorLiveData<List<PostModel>> postListMediator;
  private MediatorLiveData<SubredditModel> subMediator;

  public void init(String subredditName, SubRepository subRepo, PostRepository postRepo) {
    // ensure that the subreddit model is reinitialized when the subreddit changes
    if (!subredditName.equals(subName)) {
      Log.d(TAG, "VM to display " + subredditName + " initialized");
      subName = subredditName;
      this.postRepo = postRepo;
      this.subRepo = subRepo;

      // initialize observables
      postListMediator = new MediatorLiveData<>();
      subMediator = new MediatorLiveData<>();


      postListMediator.addSource(postRepo.getPosts(subName), new Observer<List<PostModel>>() {
        @Override
        public void onChanged(@Nullable List<PostModel> postModels) {
          if (postModels != null && postModels.isEmpty()) {
            Log.d(TAG, "empty posts");
            retrieveMorePosts(true);
          } else {
            Log.d(TAG, "not empty posts");
            postListMediator.setValue(postModels);
          }
        }
      });

      // TODO: STILL TESTING retrieve the banner image from the subredddit css
      //subRepo.getSubGateway().retrieveSubBanner(subName);
      this.subMediator.addSource(subRepo.getSingleSub(subName), new Observer<SubredditModel>() {
        @Override
        public void onChanged(@Nullable SubredditModel newModel) {
          if (newModel == null) {
            Log.d(TAG, "No subreddit saved locally, retrieving from network");
            subRepo.retrieveSingleSub(subName);
          }
          else {
            Log.d(TAG, "Subreddit loaded " + newModel.getBannerUrl());
            subMediator.setValue(newModel);
          }
        }
      });

      isInitialized = true;
    }
  }

  @Override
  public boolean isInitialized() {
    return isInitialized;
  }


  public LiveData<SubredditModel> getSubModel() {
    return subMediator;
  }


  @Override
  public String getSubName() {
    return subName;
  }


  /**
   * Exposes the livedata list of posts to the view
   * @return the lists of posts
   */
  public LiveData<List<PostModel>> getPosts() {
    return postListMediator;
  }


  /**
   * Method that allows views aware of the VM to request the VM retrieve more posts
   */
  @Override
  public void retrieveMorePosts(boolean resetPosts) {
    if (resetPosts) {
      // we pass null for the value of next to tell repo that we're refreshing
      postRepo.retrieveMorePosts(subName, null);
    }
    else {
      // retrieve the "after" value for the next posting
      postRepo.getNextPostingVal(this, subName);
    }
  }


  @Override
  public void onNextListing(String nextVal) {
    Log.d(TAG, "Retrieving next posts with " + nextVal);
    // retrieve the "after" value for the next posting
    postRepo.retrieveMorePosts(subName, nextVal);
  }




  public void updateSubStatus(boolean toSubscribe) {
    Log.d(TAG, "Changing to subscribed " + toSubscribe);

    if (toSubscribe) {
      // subscribe if not currently subscribed
      LiveData <Boolean> successObservable = subRepo.getSubGateway().subscribe(subName);
      subMediator.addSource(successObservable, (Boolean success) -> {

        if (success != null && success) {
          Log.d(TAG, "subscribing");
        }
        // unsubscribe after consuming event
        subMediator.removeSource(successObservable);
      });
    } else {
      // unsubscribe if already subscribed
      LiveData <Boolean> successObservable = subRepo.getSubGateway().unsubscribe(subName);
      subMediator.addSource(successObservable, (Boolean success) -> {

        if (success != null && success) {
          Log.d(TAG, "unsubscribing");
          //subMediator.setValue(false);
        }

        //subscribed.setValue(success);
        subMediator.removeSource(successObservable);
      });
    }
  }

}

File: app/src/main/java/com/relic/presentation/displaysub/DisplaySubVM.kt
package com.relic.presentation.displaysub

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import com.relic.data.ListingRepository
import com.relic.data.PostRepository
import com.relic.data.PostSource
import com.relic.data.SortScope
import com.relic.data.SortType
import com.relic.data.SubRepository
import com.relic.data.repository.NetworkException
import com.relic.domain.models.PostModel
import com.relic.domain.models.SubredditModel
import com.relic.network.NetworkUtil
import com.relic.presentation.base.RelicViewModel
import com.relic.presentation.main.RelicError
import com.shopify.livedataktx.SingleLiveData
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

open class DisplaySubVM (
    private val postSource: PostSource,
    private val subRepo: SubRepository,
    private val postRepo: PostRepository,
    private val postInteractor :  DisplaySubContract.PostAdapterDelegate,
    private val listingRepo : ListingRepository,
    private val networkUtil : NetworkUtil
) : RelicViewModel(), DisplaySubContract.ViewModel, DisplaySubContract.PostAdapterDelegate by postInteractor {

    class Factory @Inject constructor(
        private val subRepo: SubRepository,
        private val postRepo : PostRepository,
        private val postInteractor : DisplaySubContract.PostAdapterDelegate,
        private val listingRepo : ListingRepository,
        private val networkUtil : NetworkUtil
    ) {
        fun create (postSource : PostSource) : DisplaySubVM {
            return DisplaySubVM(postSource, subRepo, postRepo, postInteractor, listingRepo, networkUtil)
        }
    }

    private var currentSortingType = SortType.DEFAULT
    private var currentSortingScope = SortScope.NONE
    private var subPostAfter : String? = null

    private var retrievalInProgress = true

    private val _subredditMediator = MediatorLiveData<SubredditModel>()
    private val _postListMediator= MediatorLiveData<List<PostModel>> ()
    private val _subInfoLiveData = MutableLiveData<DisplaySubInfoData>()
    private val _refreshLiveData = MutableLiveData<Boolean>()
    private val _errorLiveData = SingleLiveData<RelicError>()

    val subredditLiveData : LiveData<SubredditModel> = _subredditMediator
    val postListLiveData : LiveData<List<PostModel>> = _postListMediator
    val subInfoLiveData : LiveData<DisplaySubInfoData> = _subInfoLiveData
    val refreshLiveData : LiveData<Boolean> = _refreshLiveData
    val errorLiveData : LiveData<RelicError> = _errorLiveData

    init {
        when (postSource) {
            is PostSource.Subreddit -> {
                _subredditMediator.addSource(subRepo.getSingleSub(postSource.subredditName)) {
                    _subredditMediator.setValue(it)
                }
            }
            is PostSource.Frontpage -> {}
            is PostSource.All -> {}
        }
        
        // initial check for connection -> allows us to decide if we should use retrieve posts
        // from the network or just display what we have locally
        if (networkUtil.checkConnection()) {
            retrieveMorePosts(true)
            retrieveSubreddit()
        } else {
            // observe the list of posts stored locally
            _postListMediator.addSource(postRepo.getPosts(postSource)) { postModels ->
                _postListMediator.postValue(postModels)
            }

            launch {
                subPostAfter = listingRepo.getAfter(postSource)
            }

            //subRepo.getSubGateway().retrieveSubBanner(subName);
        }

        _subInfoLiveData.postValue(DisplaySubInfoData(
            sortingMethod = currentSortingType,
            sortingScope = currentSortingScope
        ))
    }

    private fun retrieveSubreddit() {
        val subName : String?  =  when (postSource) {
            is PostSource.Subreddit -> postSource.subredditName
            is PostSource.Frontpage -> ""
            else -> null
        }

        if (subName != null) {
            // TODO: STILL TESTING retrieve the banner image from the subreddit css
            launch(Dispatchers.Main) {
                subRepo.getSubGateway().apply {
                    retrieveAdditionalSubInfo(subName)
                    retrieveSidebar(subName)
                }

                val sub = subRepo.retrieveSingleSub(subName)
                _subredditMediator.postValue(sub)
                subRepo.insertSub(sub)
            }
        }
    }

    final override fun retrieveMorePosts(resetPosts: Boolean) {
        if (networkUtil.checkConnection()) {
            launch {
                if (resetPosts) {
                    // only indicate refreshing if connected to network
                    _refreshLiveData.postValue(true)
                    val listing = postRepo.retrieveSortedPosts(postSource, currentSortingType, currentSortingScope)
                    subPostAfter = listing.data.after

                    listing.data.children?.let { posts ->
                        _postListMediator.postValue(posts)
                        // TODO add preferences manager to let us check if user wants to store the loaded posts
                        postRepo.clearAllPostsFromSource(postSource)
                        postRepo.insertPosts(postSource, posts)
                    }

                    listingRepo.insertAfter(postSource, listing.data.after)
                } else {
                    subPostAfter?.let { currentAfter ->
                        val listing = postRepo.retrieveMorePosts(postSource, currentAfter)
                        subPostAfter = listing.data.after

                        listing.data.children?.let { posts ->
                            val newPosts = _postListMediator.value!!.toMutableList()
                            newPosts.addAll(posts)

                            _postListMediator.postValue(newPosts)
                            postRepo.insertPosts(postSource, posts)
                        }

                        listingRepo.insertAfter(postSource, listing.data.after)
                    }
                }

                retrievalInProgress = false
                _refreshLiveData.postValue(false)
            }
        }
        else {
            retrievalInProgress = false
            _errorLiveData.postValue(RelicError.NetworkUnavailable)
        }
    }

    /**
     * Called when the user has changed an aspect of the sorting method for this sub. Null values
     * for either the sortType or sortScope indicate no change
     * @param sortType : code corresponding to sort type
     * @param sortScope : code corresponding to sort scope
     */
    override fun changeSortingMethod(sortType: SortType?, sortScope: SortScope?) {
        // update the current sorting method and scope if it has changed
        sortType?.let { currentSortingType = it }
        sortScope?.let { currentSortingScope = it }

        launch(Dispatchers.Main) {
            // remove all posts from current db for this subreddit (triggers retrieval)
            postRepo.clearAllPostsFromSource(postSource)
            retrieveMorePosts(true)
            _subInfoLiveData.postValue(
                DisplaySubInfoData(sortingMethod = currentSortingType, sortingScope = currentSortingScope)
            )
        }
    }

    override fun updateSubStatus(subscribe: Boolean) {
        if (postSource is PostSource.Subreddit) {
            val subName = postSource.subredditName
            Timber.d("Changing to subscribed $subscribe")

            launch(Dispatchers.Main) {
                subRepo.getSubGateway().subscribe(subscribe, subName)
            }
        }
    }

    override fun handleException(context: CoroutineContext, e : Throwable) {
        val subE = when (e) {
            is NetworkException -> RelicError.NetworkUnavailable
            else -> RelicError.Unexpected
        }
        _errorLiveData.postValue(subE)

        Timber.e(e)
    }
}
